### ISSUE TITLE
consider: `NNModule` を適用したもので、自動微分みたいなものができるのかを考えたい

### ISSUE THREAD

**Issue by gckohaku** [gckohaku/cnn-learn-rust#26]
現状は `NNModule::Linear`, `NNModule::ReLU`, `NNModule::SoftmaxAndCELoss` の組み合わせについて考えればよい
また、現時点では分岐なしのグラフとして表すことができるので、考えやすいとは思う

**Comment by gckohaku** [comment]
それぞれの関数や変数を以下のように定義
### Input
$$X$$
### Linear1, Linear2
$$f_1(X) = U_1, f_2(H_1) = U_2$$
### Linear1, Linear2 内の重み
$$W_1, W_2$$
### ReLU1
$$a_1(U_1) = H_1$$
### Softmax
$$\sigma (U_2) = Y$$
### CrossEntropyLoss
$$c(Y) = l$$
### ターゲット値
$$T$$

それぞれのモジュールでの偏微分は以下のようになる
### 出力層の勾配
$$\frac{\partial l}{\partial U_2}=Y-T$$
### Linear2 の勾配
$$\frac{\partial U_2}{\partial H_1}=(W_2)^T$$
### ReLU1 の勾配
$$\frac{\partial H_1}{\partial U_1}=a_1 \prime(U_1)$$
### Linear1 の勾配
$$\frac{\partial U_1}{\partial X}=(W_1)^T$$

### Linear2 での重み、バイアスへの勾配
$$\frac{\partial U_2}{\partial W_2}=(H_1)^T$$
$$\frac{\partial U_2}{\partial B_2}=I\quad\text{(単位行列)}$$
### Linear1 での重み、バイアスへの勾配
$$\frac{\partial U_1}{\partial W_1}=X^T$$
$$\frac{\partial U_1}{\partial B_1}=I$$

**Comment by gckohaku** [comment]
動的ディスパッチ、enum ディスパッチの両方で完了した
